WIP: Sensitivity calculation updates and working script

misabelber opened the pull request:
Hi all, 
I've been working over Ruben's code and arranged a script that calculates the sensitivity optimizing cuts in theta2 and gammaness. It basically does the same thing that the function  sensitivity() from lstchain.mc.sensitivity, but with some changes and hardcoded stuff I've been trying.
The script test_sensitivity.py is commented to be understandable, so if you want to run it you can follow what is being done. 
I've added an if statement in the calculation of theta2, so for protons their src position is taken in the center.
Also I've changed the range of theta2, from 0.05 to 0.1 because some energy bins were empty.

There are still things that need to be worked on:
- Get right the cuts in theta2 for protons
- Get right the number of off regions
- Requirements on min. number of events for sensitivity calc

The resuling sensitivity curve obtained right now looks like this:
![imagen](https://user-images.githubusercontent.com/38687749/57860507-45029200-77f5-11e9-9014-7070b93a33cc.png)

vuillaut commented:
Hi @misabelber.

Nice to see the first sensitivity plot!

Please merge master into your branch and fix tests.
I'll have a look into this asap.

rlopezcoto commented on scripts/test_sensitivity.py:
```diff
@@ -0,0 +1,171 @@
+            if np.isnan(sens[i][j][k]) or np.isinf(sens[i][j][k]) or sens[i][j][k]==0:
+                sens[i][j][k] = 1e100
+            if pre_gamma[i][j][k] < min_num_events or pre_hadron[i][j][k] < min_num_events:
+                sens[i][j][k] = 1e100
```
this condition should be applied to `final_gamma[i][j][k]` and `final_hadrons[i][j][k]`

rlopezcoto replied:
you can also add the condition that `final_gamma[i][j][k] > final_hadrons[i][j][k] * 0.05` (the excesses need to be > 5% of the background)

misabelber commented on lstchain/spectra/crab.py:
```diff
@@ -46,19 +46,19 @@ def crab_hegra(E):
     dFdE  = f0 * np.power(E / e0, alpha)
 
-    return dFdE.to(1/u.TeV / u.cm**2 / u.s), par
+    return dFdE.to(1/u.GeV / u.cm**2 / u.s), par
```
There was a problem because in some places energies were in TeV (spectrum of crab, protons...) and in others (energies from dl2 files, limits of energy bins...), energies were in GeV. That was one of the main reasons from the scaling problems of the curves I showed you at first. I decided to put all energies of the code in GeV, and only transform to TeV at the time to plot the sensitivity curve.

vuillaut replied:
Long-standing issue, related to #132

misabelber commented on scripts/test_sensitivity.py:
```diff
@@ -0,0 +1,171 @@
+
+            final_gamma[i][j][k] = eg_w_sum * obstime.to_value()
+            final_hadrons[i][j][k] = ep_w_sum * obstime.to_value()
+
```
The problem here is that eg_w_sum and ep_w_sum come from the sum of a "Series" object (from the pandas dataframe (e_reco_gw, e_reco_pw)) which doesn't allow operations with units (actually they are multiplied by the weights which units are 1/s, but this unit doesn't appear afterwards) so even the units being correct in the steps before, Python raises an error here when combining unitless quantities with the obstime units. I haven't found an easy way to keep the units from e_reco_gw, e_reco_pw but I did all the checks to be sure that all units are correct. They can be added manually though.

vuillaut replied:
Then you can go the other way around, no?
Transform your pandas object in numpy array before multiplying by arrays with units. This way, you will keep track of the unit.

misabelber replied:
Yep, that can be a solution! I'm on it :)

misabelber commented on scripts/test_sensitivity.py:
```diff
@@ -0,0 +1,171 @@
+#Calculate the number of off positions
+noff=mc_par_p['area_sim']/mc_par_g['area_sim']
+#@MAB: I think this is wrong but I'm not sure how to calculate it...
+
```
Ok! Then I'll set it to 1 for now :')

misabelber commented:
Hi! 
I applied the requested changes:
 - Pandas object is transformed to numpy array so units are conserved
- noff = 1
- theta2max back to 0.05
- apply condition of max events to final_gamma, final_hadron.
- add condition of excess > 5% background

Also, removed script test_sensitvity.py, moved everything to function lstchain.mc.sensitivity.sens() and added script lst-sensitivity.py where the parser arguments are the simtelfiles and dl2 files. 

There is one feature I'm not using right now which is the function sens_minimization_plot() because it's throwing an error: TypeError("Invalid dimensions for image data") and I haven't been able to solve it yet. Maybe @rlopezcoto can know better.

rlopezcoto commented:
Great! 
Regarding the error, if you have 12 bins in energy and 10 in th2 and gammaness, it should not be happening...
If the plot is not present, I would at least add some `print` for the number of gammas/hadrons that are used for the calculation of the sensitivity in each energy bin (after you have selected the minimum, so one print per energy bin).
I would accept after adding this, so we can keep working on this (still not sure whether the background region is properly selected...)

misabelber commented:
Yep, I'm giving 12 energy bins and 10 for gammaness and theta2 but still. I must say that I made that function work before PR #105 got merged so I'll try to find what changes I did that somehow got lost.
 I can add that print for now.
What do you mean with the background region being properly selected? If you mean the thing about applying different theta2 cuts for protons, (allowing them only in a region around 0.4º), that is not included yet. I'm not sure how to do that cut,  but I guess maybe before doing any calculation, before weighting, etc, just throw away the protons outside that region?

vuillaut commented on lstchain/mc/sensitivity.py:
```diff
@@ -68,35 +70,42 @@ def process_mc(simtelfile, dl2_file):
-    theta2 = (events.src_x - events.src_x_rec)**2 + \
-        (events.src_y - events.src_y_rec)**2 * u.deg**2
+
+    if events.iloc[0].mc_type==0:
```
I don't quite understand this condition.
- only the type of the first event is checked - shouldn't one filter all the events to keep, .e.g `events[events.mc_type==0]` ?
- if not a gamma, why theta2 is computed relative to the camera center?

misabelber replied:
- Only the first event is checked because as it is coded now the files that you give should contain only events from one type of particle (a file of gammas or a file of protons), but yes I guess they can be filtered  in case at some point we give mixed files. 

- At first I was using the same theta2 definition, but @moralejo and @rlopezcoto explained that we need to calculate the rate of protons in a region close to the poisition of the gamma source.

vuillaut replied:
- It's not more complicated to filter and will avoid future issues. If you could change that please.
- still don't get it, if you want the rate of protons close to the position of the gamma source, you should also compute theta2 relative to `events.src_x[events.mc_type==0]`, no ?

misabelber replied:
I'm taking a look at the filtering thing, but I don't see it clear, because the function is constructed in a way it's assumed that you are going to pass to it only events of one type (since it takes a simtel file) and if we apply a filter, it means that you could pass events of two types and then return two sets of theta2 values, and one of them won't be related to the simtelfile... I think is a bit messy. Anyway I think this function will be changed once we have the new DL1b format in which the mc info is stored in the .h5 file so it won't be necessary to read the simtelfile.

vuillaut replied:
Yes, I think the confusion comes from the fact of using a single function to make two very different processing.
The theta2 calculated here for gammas is of a different nature than the one calculated for hadrons (and electrons?) and I got confused by it.
Maybe we should consider breaking it down to different functions.

vuillaut replied:
There is a `ring_theta2` and a `source_theta2`.
Then they are applied to different particle types.

rlopezcoto replied:
I had a chat with @vuillaut about something else and we also talked about this... probably the best would be to propagate the information of `events.src_x` and `events.src_x_rec` (also the y coordinate) further, so we can compute later the theta2 for gammas and the angular distance for protons. We can program a ring for the proton (electron) extraction later, is that ok?

rlopezcoto replied:
Related to this... `events.src_x` has units of `m`, right? If it is like this, then the theta2 is completely wrong...

vuillaut commented on lstchain/mc/sensitivity.py:
```diff
@@ -106,86 +115,138 @@ def bin_definition(gb, tb):
+    for i in range(0, eb):
+        for j in range(0, gb):
+            for k in range(0, tb):
+                if np.isnan(sens[i][j][k]) or np.isinf(sens[i][j][k]) \
```
Can be simplified imo with 
```
conditions = (not np.isfinite(sens[i,j,k])) or (sens[i,j,k]<=0) or (final_gamma[i,j,k] < min_num_events) or (final_hadrons[i,j,k] < min_num_events) or (not final_gamma[i,j,k] > final_hadrons[i,j,k] * 0.05)
if cond:
    sens[i][j][k] = 1e100
```
why 1e100 and not `np.nan` stating clearly this is not a good value ?

misabelber replied:
I set it to a huge value because later, the loop to chose the bin with minimum sensitivity was getting crazy with nans and infs.

vuillaut replied:
In this case, the isfinite condition should be verified in the loop imo.

misabelber replied:
Finally got to change np.nan for np.inf because np.argmin() was taking "nan" as a minimum, and np.nanarming() doesn't like when an array is full of nans...

vuillaut replied:
I would personally prefer to have `nan` values when the sensitivity has not been calculated properly but if it's working like that, I think we can go with it.

vuillaut commented on lstchain/mc/sensitivity.py:
```diff
@@ -74,6 +72,8 @@ def process_mc(simtelfile, dl2_file):
 
+    gevents = events[events.mc_type==0]
+
     if events.iloc[0].mc_type==0:
```
If you apply filter, then you want to work on `gevents` no? 😄

misabelber replied:
Ups!, this remained from my tests on filtering, shouldn't have been in the commit, just removed it :')

vuillaut replied:
However, the code is still checking the first event only and therefore assumes that all particles are from the same type.
If you want to force the fact that files are from a single particle type, maybe add an assert about it somewhere. I feel like this kind of implicit assumption is error-prone for the future.

misabelber replied:
I just included a way of filtering events, adding the type of particle as a function argument, so we can be sure that the resulting parameters belong to the desired type.

rlopezcoto commented:
To solve the issue of the protons and not proper sensitivity calculation, I implemented:

- the function `ring_containment`, that calculates the number of protons contained inside a ring and returns a boolean array and the area of that ring. The number of protons is normalized to the areas defined by the different theta2 cuts. As suggested by @moralejo, I multiply the result after normalization by `noff` to mimic the behavior of using a given number of OFF positions.

- Li & Ma significance calculation using `gammapy.stats.poisson`

The thing missing on this PR is to confirm or not that `events.src_x` and `events.src_x_rec` are in `m` or `deg`. If they are in units of `m`, we need to perform a coordinate transformation before applying any of the above calculations.

@misabelber could you please check if everything works alright with the implemented changes? If it does, you can change `calculate_sensitivity` by `calculate_sensitivity_lima` and if everything is alright, the results should be slightly worse. You can also check changing `noff=5` to see if results are slightly better than `noff=1`
I'll also try to add test modules.

vuillaut commented:
> The thing missing on this PR is to confirm or not that `events.src_x` and `events.src_x_rec` are in `m` or `deg`. If they are in units of `m`, we need to perform a coordinate transformation before applying any of the above calculations.

@rlopezcoto meters.

misabelber commented:
Hi all!
I've checked latests additions from @rlopezcoto and it seems to work fine, below there are some results.
I can confirm that src_x, src_x_rec are in meters so we should then do the transformation to AltAz, right?

Old Sensitivity calculation :
![imagen](https://user-images.githubusercontent.com/38687749/58006233-e20c4600-7ae7-11e9-9a0d-6c33cfff755a.png)
BEST CUTS: 
Energy bin(GeV) Gammaness Theta2(deg)
5.00 0.00 0.01
10.77 0.00 0.01
23.21 0.22 0.02
50.00 0.56 0.01
107.72 0.78 0.01
232.08 0.89 0.01
500.00 0.89 0.01
1077.22 0.89 0.01
2320.79 0.89 0.01
5000.00 0.78 0.01
10772.17 0.56 0.03
23207.94 0.44 0.03

Sensitivity LiMa:
![imagen](https://user-images.githubusercontent.com/38687749/58006289-0a944000-7ae8-11e9-8a58-95a41b29a518.png)
BEST CUTS
Energy bin(GeV) Gammaness Theta2(deg)
5.00 0.00 0.01
10.77 0.00 0.01
23.21 0.22 0.04
50.00 0.56 0.01
107.72 0.78 0.01
232.08 0.78 0.01
500.00 0.89 0.01
1077.22 0.89 0.01
2320.79 0.89 0.01
5000.00 0.78 0.01
10772.17 0.56 0.03
23207.94 0.44 0.03


LiMa, noff=5
![imagen](https://user-images.githubusercontent.com/38687749/58006349-2a2b6880-7ae8-11e9-8512-ed49da546a0c.png)
BEST CUTS: 
Energy bin(GeV) Gammaness Theta2(deg)
5.00 0.00 0.01
10.77 0.00 0.01
23.21 0.22 0.04
50.00 0.56 0.01
107.72 0.78 0.01
232.08 0.89 0.01
500.00 0.89 0.01
1077.22 0.89 0.01
2320.79 0.89 0.01
5000.00 0.78 0.01
10772.17 0.56 0.03
23207.94 0.44 0.03

rlopezcoto commented:
Thanks @misabelber! Yes, we should transform to AltAz to get `delta_az,delta_az`. For this you need the focal length and pointing position of the array.
Whenever we change the source position in camera coordinates by the angular one, the calculation should be correct

vuillaut commented:
You may use the function `camera_to_sky` for that transformation.

moralejo commented:
On Mon, 20 May 2019 at 10:57, Thomas Vuillaume <notifications@github.com>
wrote:

> You may use the function camera_to_sky for that transformation.
>
Is the transformation to Alt Az coordinates needed?
In order to calculate theta2 we need to obtain, from the gamma direction
(in alt az), the corresponding position on the camera (src_x, src_y,
correct?), making sure of using the *effective*  focal length, to account
for the spot size shift due to aberration. The latter is relevant mostly
for off-axis observations.
Then theta2 can be calculated on the plane of the camera using the linear
units (meters for example), and convert the result to angle^2 using the
deg/m plate scale of the telescope (calculated again with the effective
focal length). That should be accurate enough.




-- 

Abelardo Moralejo Olaizola
Institut de Física d'Altes Energies
Tel : +34 931641662
Fax: +34 935811938
Avís - Aviso - Legal Notice - (LOPD) - http://legal.ifae.es

-- 
Avís -
Aviso - Legal Notice - (LOPD) - http://legal.ifae.es 
<http://legal.ifae.es/>

misabelber commented:
Hi!
I just included the transformation to angular coordinates in the function process_mc(). The curves look prettier now:

Li&Ma, noff=1
![imagen](https://user-images.githubusercontent.com/38687749/58014602-834fc800-7af9-11e9-815a-28f53b66ef95.png)
BEST CUTS: 
Energy bin(GeV) Gammaness Theta2(deg2)
5.00 0.00 0.01
10.77 0.00 0.01
23.21 0.11 0.05
50.00 0.22 0.04
107.72 0.56 0.04
232.08 0.78 0.02
500.00 0.89 0.02
1077.22 0.78 0.01
2320.79 0.78 0.02
5000.00 0.22 0.03
10772.17 0.11 0.04
23207.94 0.00 0.01

Li&Ma, noff=5
![imagen](https://user-images.githubusercontent.com/38687749/58014731-c5790980-7af9-11e9-81ab-445e366c859b.png)
BEST CUTS: 
Energy bin(GeV) Gammaness Theta2(deg)
5.00 0.00 0.01
10.77 0.00 0.01
23.21 0.11 0.05
50.00 0.22 0.04
107.72 0.56 0.04
232.08 0.78 0.02
500.00 0.89 0.01
1077.22 0.78 0.01
2320.79 0.78 0.02
5000.00 0.22 0.02
10772.17 0.11 0.04
23207.94 0.00 0.01

misabelber commented:
Travis tests are failing because I've used ctaplot to calculate theta2, shall we:
- Make ctaplot a requirement for lstchain?
- Include the calculation of theta2 somwhere in lstchain? (in reco.utils for example)

vuillaut commented:
> Travis tests are failing because I've used ctaplot to calculate theta2, shall we:
> 
> * Make ctaplot a requirement for lstchain?
> * Include the calculation of theta2 somwhere in lstchain? (in reco.utils for example)

My intention was to use ctaplot in the future for all the reconstruction plots anyway.
I don't want to impose it of course but if you are happy about it, it will certainly ease the validation processes.

rlopezcoto commented:
To be honest, I do not see necessary to include an external dependency of a non-standard library only for a theta2 calculation that can be easily performed in one line of code.

misabelber commented:
Well actually we could consider ctaplot as a substitution of the lstchain module "visualization", so it won't only be for the theta2, but for the plotting of all the reconstruction results.

vuillaut commented:
> Well actually we could consider ctaplot as a substitution of the lstchain module "visualization", so it won't only be for the theta2, but for the plotting of all the reconstruction results.

That's what I meant.
I think @misabelber is already using it (as I do) for most of the presented result plots.
And the `lstchain` visualization module would benefit from some refactoring anyway...

rlopezcoto commented:
As I said, I do not see it necessary for the `theta2` calculation and would rather not include any dependencies from personal packages into cta-lstchain

vuillaut commented:
For the angular separation, you may use
`astropy.coordinates.angle_utilities.angular_separation`

rlopezcoto commented:
> BEST CUTS:
Energy bin(GeV) Gammaness Theta2(deg)
5.00 0.00 0.01
10.77 0.00 0.01
23.21 0.11 0.05
50.00 0.22 0.04
107.72 0.56 0.04
232.08 0.78 0.02
500.00 0.89 0.01
1077.22 0.78 0.01
2320.79 0.78 0.02
5000.00 0.22 0.02
10772.17 0.11 0.04
23207.94 0.00 0.01

Thanks @misabelber for the results! Could you please also include in the printing the number of protons/gammas that you are using for the calculation in each energy bin? 
Some cuts look rather large in theta2 (for example `theta2`=0.04 for `Energy`=50, 107 GeV), with very low gamma-hadron separation (same for high energies!).

- After this, the addition of electrons should not be difficult and it would help to give a more realistic estimation of the sensitivity
- Some more refinement: in CTA the sensitivity is defined using 5 bins per decade, implying a lower number of background events and slightly worst sensitivity
(these two last points can wait and otherwise we could merge this PR after adding this printing I mentioned)

misabelber commented:
Thanks @rlopezcoto ! Yes sure, I'll work on that printing rightaway. Just changed the angular separation calculation to use astropy function as @vuillaut suggested.

misabelber commented:
I just included the number of gammas and protons in the print. 
Using for example 21 bins in energy which are ~5 bins per decade the results are:
![imagen](https://user-images.githubusercontent.com/38687749/58031237-4ea33700-7b20-11e9-90d8-c5c3551e44f2.png)
BEST CUTS: 
Energy bin(GeV) Gammaness Theta2(deg2) #Gammas #Protons
5.00 0.00 0.01 0.00 0.00
7.75 0.00 0.01 0.00 0.00
12.02 0.00 0.01 0.00 0.00
18.64 0.00 0.01 11.97 504.42
28.90 0.11 0.05 315.95 1715.85
44.81 0.22 0.05 3044.68 1165.37
69.47 0.11 0.04 12167.33 712.04
107.72 0.44 0.04 14501.63 433.65
167.02 0.56 0.03 11703.75 300.13
258.97 0.67 0.03 8320.46 192.13
401.54 0.78 0.02 5062.33 123.40
622.60 0.89 0.02 2457.75 55.59
965.35 0.89 0.01 1160.11 34.92
1496.79 0.56 0.02 847.30 33.08
2320.79 0.78 0.02 353.99 23.38
3598.43 0.22 0.02 197.04 24.89
5579.42 0.11 0.03 104.20 25.47
8650.98 0.11 0.03 39.94 32.47
13413.48 0.00 0.04 16.62 16.35
20797.81 0.00 0.05 10.02 70.84
32247.33 0.00 0.01 0.68 3.94

rlopezcoto commented:
it looks good to me now, thanks! (already approved)
what happened to the condition of >10 events in the last bin?
> 32247.33 0.00 0.01 0.68 3.94

moralejo commented:
Thanks for the update. Although the shape of the curve makes sense (and it
is nice to see its smoothness from large statistics), the values are still
too good for a single telescope. This is the average of north- and
south-wise pointings, right? At 200 GeV you have ~0.9e-12 TeV cm-2 s-1,
i.e. 1.4e-12 erg cm-2 s-1. I checked the value for 2 LSTs in the latter
units, and it is 2e-12 (MARS analysis).
You can also check the public CTA sensitivity curves, where VERITAS, MAGIC
and HESS are shown for comparison, and they are all at 200 GeV above
1.4e-12. There is no way a single LST can beat them at that energy. There
must be still some problem in the calculations, I am afraid.

On Mon, 20 May 2019 at 17:00, Mab Bernardos <notifications@github.com>
wrote:

> I just included the number of gammas and protons in the print.
> Using for example 21 bins in energy which are ~5 bins per decade the
> results are:
> [image: imagen]
> <https://user-images.githubusercontent.com/38687749/58031237-4ea33700-7b20-11e9-90d8-c5c3551e44f2.png>
> BEST CUTS:
> Energy bin(GeV) Gammaness Theta2(deg2) #Gammas #Protons
> 5.00 0.00 0.01 0.00 0.00
> 7.75 0.00 0.01 0.00 0.00
> 12.02 0.00 0.01 0.00 0.00
> 18.64 0.00 0.01 11.97 504.42
> 28.90 0.11 0.05 315.95 1715.85
> 44.81 0.22 0.05 3044.68 1165.37
> 69.47 0.11 0.04 12167.33 712.04
> 107.72 0.44 0.04 14501.63 433.65
> 167.02 0.56 0.03 11703.75 300.13
> 258.97 0.67 0.03 8320.46 192.13
> 401.54 0.78 0.02 5062.33 123.40
> 622.60 0.89 0.02 2457.75 55.59
> 965.35 0.89 0.01 1160.11 34.92
> 1496.79 0.56 0.02 847.30 33.08
> 2320.79 0.78 0.02 353.99 23.38
> 3598.43 0.22 0.02 197.04 24.89
> 5579.42 0.11 0.03 104.20 25.47
> 8650.98 0.11 0.03 39.94 32.47
> 13413.48 0.00 0.04 16.62 16.35
> 20797.81 0.00 0.05 10.02 70.84
> 32247.33 0.00 0.01 0.68 3.94
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cta-observatory/cta-lstchain/pull/131?email_source=notifications&email_token=ADMCCOXJMDGE43YZEBPGOHLPWK4JJA5CNFSM4HNM7CK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVZDPKY#issuecomment-494024619>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADMCCORB5QCEP4T3JTTEFB3PWK4JJANCNFSM4HNM7CKQ>
> .
>


-- 

Abelardo Moralejo Olaizola
Institut de Física d'Altes Energies
Tel : +34 931641662
Fax: +34 935811938
Avís - Aviso - Legal Notice - (LOPD) - http://legal.ifae.es

-- 
Avís -
Aviso - Legal Notice - (LOPD) - http://legal.ifae.es 
<http://legal.ifae.es/>

moralejo commented:
By the way, would be nice to show a better parametrization of the Crab spectrum (like a log-parabola), the power-law is overshooting the actual spectrum significantly at the lower energies.

rlopezcoto commented:
Hi @moralejo, I was also a bit worried about the values, but was hopping that the reason was that electrons are not included. 
Also, I'm not sure these results are for one or two pointings, @misabelber can comment. I can implement the log-parabola, but what was used for the reweighting on the CTA sensitivity curves?

rlopezcoto commented:
Or is it only for the comparison lines? That is already implemented (`crab_HEGRA -> crab_MAGIC`)

moralejo commented:
On Mon, 20 May 2019 at 23:24, Ruben Lopez-Coto <notifications@github.com>
wrote:

> Hi @moralejo <https://github.com/moralejo>, I was also a bit worried
> about the values, but was hopping that the reason was that electrons are
> not included.
>
That is relevant indeed, but I doubt very much it can explain it all. The
difference single <---> stereo performance should be way larger than the
effect of electrons at that energy (especially for a single telescope,
which does not kill protons so efficiently).

Also, I'm not sure these results are for one or two pointings, @misabelber
> <https://github.com/misabelber> can comment. I can implement the
> log-parabola, but what was used for the reweighting on the CTA sensitivity
> curves?
>
What you use for the gamma weighting in the calculations does not matter so
much, because it only affects the relative weights of events *within a bin*
- and bins are quite narrow, anyway (and the difference between "true" and
"power-law" indices will be <0.6 or so). So the points would hardly change
(in absolute flux units).

-- 

Abelardo Moralejo Olaizola
Institut de Física d'Altes Energies
Tel : +34 931641662
Fax: +34 935811938
Avís - Aviso - Legal Notice - (LOPD) - http://legal.ifae.es

-- 
Avís -
Aviso - Legal Notice - (LOPD) - http://legal.ifae.es 
<http://legal.ifae.es/>

rlopezcoto commented:
I changed `crab_hegra` by `crab_magic` for the plotting. 
It seems that we have not converged yet on the results being correct or not, but I would merge this PR and investigate this in parallel, is that ok @moralejo, @vuillaut, @misabelber?

misabelber commented:
Hi all!
@rlopezcoto  About bins with less than 10 events, even they are being printed, the sensitivity is not calculated for them (is setlled to np.inf) so those bins doesn't appear in the sensitivity curve plot.

@moralejo Those results are only for South pointing (gammas and protons). 

I will now start  doing checks to see where can be the problem that is giving too good results. 
Thank you!

rlopezcoto commented:
@vuillaut is it ok is I merge it so we can keep investigating?

vuillaut commented:
There is 0 unit test attached to these (many, important) additional functions.
Asking for unit tests for every function is most probably overkill here but I would very much like to see a couple of unit tests (at least one global) doing the DL2 to DL3 steps that @misabelber must have already implemented in a script to make the curve. Just to make sure the chaining of functions works and prevent future big breaks.
I suggest that the unit test `apply_models` actually dump a DL2 file on disk and that this one is loaded by a `test_sensitivity_calculation` for example.

rlopezcoto commented:
Hi,
I prepared unit tests for all the sensitivity functions, will do the PR when this one gets merged. For the DL2 parameter calculation I agree we should have some, maybe an issue can be opened?

rlopezcoto commented:
Hi,
I prepared unit tests for all the sensitivity functions, will do the PR when this one gets merged. For the DL2 parameter calculation I agree we should have some, maybe an issue can be opened?

misabelber commented on lstchain/mc/sensitivity.py:
```diff
@@ -68,124 +73,279 @@ def process_mc(simtelfile, dl2_file):
+    alt1 = src_pos_reco.alt.rad
+    az1 = np.arctan(np.tan(src_pos_reco.az.rad))
+
+    angdist2 = (angular_separation(az1, alt1, az2, alt2).to_numpy() * u.rad)**2
```
Yes, it works because actually alt1, az1, alt2, az2 are Series objects (from the pandas dataframe) so the result from angular_separation won't conserve any unit.

vuillaut replied:
Ho yes you are right, my bad.

vuillaut commented:
Ok.
The DL2 parameter calculation is already done, they only need to be saved on disk for next step.
Ok let's move on with this, please add individual and global unit tests asap.

vuillaut commented:
> I just included the number of gammas and protons in the print.
> Using for example 21 bins in energy which are ~5 bins per decade the results are:
> ![imagen](https://user-images.githubusercontent.com/38687749/58031237-4ea33700-7b20-11e9-90d8-c5c3551e44f2.png)

@misabelber 
For the sensitivity curve, did you filter on telescope id to have only LST1?
Otherwise, the sensitivity is overestimated by a factor ~4.

misabelber commented:
@vuillaut Ah, I think not, I'll check it out. Thanks!
